File: sdk/python/src/waeasi/runtime/headers.py
"""waeasi.runtime.headers — RFC-compliant header container.

Insertion order is preserved (required for HTTP/2 pseudo-header
positioning); names are case-insensitive on lookup but stored
lower-cased for fast equality.

Multi-value semantics match WHATWG Fetch: ``append`` keeps duplicates
in order; ``get`` returns a comma-joined string; ``get_all`` exposes
the raw list.
"""

from __future__ import annotations

import re
from typing import Iterable, Iterator, Mapping, Optional, Union

_TOKEN = re.compile(r"^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$")
_BAD = re.compile(r"[\r\n\0]")

HeadersInit = Union[
    "Headers",
    Mapping[str, Union[str, Iterable[str]]],
    Iterable[tuple[str, str]],
    None,
]


class Headers:
    __slots__ = ("_map",)

    def __init__(self, init: HeadersInit = None) -> None:
        self._map: dict[str, list[str]] = {}
        if init is None:
            return
        if isinstance(init, Headers):
            for k, v in init.entries():
                self.append(k, v)
            return
        if isinstance(init, Mapping):
            for k, v in init.items():
                if isinstance(v, str):
                    self.append(k, v)
                else:
                    for vv in v:
                        self.append(k, vv)
            return
        for k, v in init:
            self.append(k, v)

    @staticmethod
    def _norm(name: str) -> str:
        if not _TOKEN.match(name):
            raise ValueError(f"invalid header name: {name!r}")
        return name.lower()

    @staticmethod
    def _check(value: str) -> str:
        if _BAD.search(value):
            raise ValueError("header value contains forbidden char")
        return value

    def append(self, name: str, value: str) -> None:
        k = self._norm(name)
        v = self._check(value)
        self._map.setdefault(k, []).append(v)

    def set(self, name: str, value: str) -> None:
        self._map[self._norm(name)] = [self._check(value)]

    def get(self, name: str) -> Optional[str]:
        v = self._map.get(self._norm(name))
        return ", ".join(v) if v else None

    def get_all(self, name: str) -> list[str]:
        return list(self._map.get(self._norm(name), ()))

    def has(self, name: str) -> bool:
        return self._norm(name) in self._map

    def delete(self, name: str) -> None:
        self._map.pop(self._norm(name), None)

    def entries(self) -> Iterator[tuple[str, str]]:
        for k, vs in self._map.items():
            for v in vs:
                yield k, v

    def keys(self) -> Iterator[str]:
        yield from self._map.keys()

    def values(self) -> Iterator[str]:
        for vs in self._map.values():
            yield from vs

    def __iter__(self) -> Iterator[tuple[str, str]]:
        return self.entries()

    def __len__(self) -> int:
        return sum(len(v) for v in self._map.values())

    def to_wasi(self) -> list[tuple[str, bytes]]:
        return [(k, v.encode("utf-8")) for k, v in self.entries()]

    @classmethod
    def from_wasi(cls, pairs: Iterable[tuple[str, bytes]]) -> "Headers":
        h = cls()
        for k, v in pairs:
            h.append(k, v.decode("utf-8", errors="replace"))
        return h

    def __repr__(self) -> str:
        return f"Headers({list(self.entries())!r})"

File: sdk/typescript/src/runtime/headers.ts
// @waeasi/sdk — Headers
//
// RFC-compliant header container used by both Request and Response.
// The implementation is intentionally allocation-light: header names
// are normalized once on insert, lookup is O(1), and iteration order
// is insertion-stable (required for HTTP/2 pseudo-header positioning).
//
// Only ASCII is allowed in names; values are stored as UTF-8 bytes so
// we can pass them straight to wasi:http/types without re-encoding.

const TOKEN_RE = /^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/;
const NAME_BAD = /[\r\n\0]/;
const VAL_BAD = /[\r\n\0]/;

export type HeadersInit =
    | Headers
    | Record<string, string | string[]>
    | Array<[string, string]>;

export class Headers {
    /** Lower-cased name → array of values (multi-value preserves order). */
    private readonly map: Map<string, string[]> = new Map();

    constructor(init?: HeadersInit) {
        if (!init) return;
        if (init instanceof Headers) {
            for (const [k, v] of init.entries()) this.append(k, v);
            return;
        }
        if (Array.isArray(init)) {
            for (const [k, v] of init) this.append(k, v);
            return;
        }
        for (const k of Object.keys(init)) {
            const v = (init as Record<string, string | string[]>)[k];
            if (Array.isArray(v)) for (const vv of v) this.append(k, vv);
            else this.append(k, v);
        }
    }

    private static norm(name: string): string {
        if (!TOKEN_RE.test(name)) {
            throw new TypeError(`invalid header name: ${name}`);
        }
        return name.toLowerCase();
    }

    private static check(value: string): string {
        if (NAME_BAD.test(value) || VAL_BAD.test(value)) {
            throw new TypeError("header value contains forbidden char");
        }
        return value;
    }

    append(name: string, value: string): void {
        const k = Headers.norm(name);
        const v = Headers.check(value);
        const cur = this.map.get(k);
        if (cur) cur.push(v);
        else this.map.set(k, [v]);
    }

    set(name: string, value: string): void {
        this.map.set(Headers.norm(name), [Headers.check(value)]);
    }

    get(name: string): string | null {
        const v = this.map.get(Headers.norm(name));
        return v ? v.join(", ") : null;
    }

    getAll(name: string): string[] {
        return this.map.get(Headers.norm(name))?.slice() ?? [];
    }

    has(name: string): boolean {
        return this.map.has(Headers.norm(name));
    }

    delete(name: string): void {
        this.map.delete(Headers.norm(name));
    }

    *entries(): IterableIterator<[string, string]> {
        for (const [k, vs] of this.map) for (const v of vs) yield [k, v];
    }

    *keys(): IterableIterator<string> {
        for (const k of this.map.keys()) yield k;
    }

    *values(): IterableIterator<string> {
        for (const vs of this.map.values()) for (const v of vs) yield v;
    }

    [Symbol.iterator](): IterableIterator<[string, string]> {
        return this.entries();
    }

    /** Marshal to the (name, value) pair list expected by wasi:http. */
    toWasi(): Array<[string, Uint8Array]> {
        const enc = new TextEncoder();
        const out: Array<[string, Uint8Array]> = [];
        for (const [k, v] of this.entries()) out.push([k, enc.encode(v)]);
        return out;
    }

    /** Hydrate from wasi:http (name, value-bytes) pairs. */
    static fromWasi(pairs: Array<[string, Uint8Array]>): Headers {
        const dec = new TextDecoder("utf-8", { fatal: false });
        const h = new Headers();
        for (const [k, v] of pairs) h.append(k, dec.decode(v));
        return h;
    }
}
